src/NUnitFramework/tests/Constraints/NUnitEqualityComparerTests.cs

```diff
@@ -306,6 +308,89 @@ public void IEnumeratorIsDisposed()
             Assert.True(comparer.AreEqual(enumeration, enumeration, ref tolerance));
             Assert.That(enumeration.EnumeratorsDisposed);
         }
+
+        [TestCaseSource(nameof(GetRecursiveContainsTestCases))]
+        public void SelfContainedItemFoundInCollection<T>(T x, ICollection y)
+        {
+            var equalityComparer = new NUnitEqualityComparer();
+            var tolerance = Tolerance.Default;
+            var equality = equalityComparer.AreEqual(x, y, ref tolerance);
+
+            Assert.IsFalse(equality);
+            Assert.Contains(x, y);
+            Assert.That(y, Contains.Item(x));
+            Assert.That(y, Does.Contain(x));
+        }
+
+        [TestCaseSource(nameof(GetRecursiveComparerTestCases))]
+        public void SelfContainedItemDoesntRecurseForever<T>(T x, ICollection y)
+        {
+            var equalityComparer = new NUnitEqualityComparer();
+            var tolerance = Tolerance.Default;
+            equalityComparer.ExternalComparers.Add(new DetectRecursionComparer(30));
```

Comment:
A naive + simple way to detect stack overflows when unit testing the `NUnitEqualityComparer` directly. For each call to the same instance of `NUnitEqualityComparer`, use this external comparer to hook into a check of the current execution stack depth.